### ISSUE TITLE
DPDK Support

### ISSUE THREAD

**Issue by PatrykDiak** [alauda/kube-ovn#104]
Hello,

What is the timeline for DPDK support?
Is it something that you plan to implement in the near future?

Thanks,
Patryk

**Comment by halfcrazy** [comment]
Hello,
My colleague is working on the Mellanox hardware offload, and maybe release in Oct or Nov. For DPDK I can't promise accurate time.

**Comment by sbskas** [comment]
We are now mid-november. What is the status of this feature ?

**Comment by oilbeater** [comment]
@sbskas we still lack professionals in this field to make some progress and we are seeking help from community

**Comment by sbskas** [comment]
@oilbeater, as far as I understand, DPDK support begin by those tasks :
1. install dpdk libraries
2. install openvswitch dpdk enabled libraries
3. prepare the  [requirements for the host / pod / vm](https://docs.openvswitch.org/en/latest/intro/install/dpdk/)
4. modifiy the code of ovn-kube to be aware of dpdk

Points 1-2 should be easily reachable.
Point 3 is unknown. One has to be aware of 2 cases : VMware with vmxnet3 / Bare Metal host with compatible hardware. I don't really know if the hugepage settings / bios tweaks are relevant for a VM.
Point 4 is unknown.

**Comment by oilbeater** [comment]
Thanks @sbskas, we will take you advice and do some investigation

**Comment by oilbeater** [comment]
Some preliminary investigations：
1. To use ovs-dpdk we need build DPDK on host and rebuild ovs with corresponding DPDK flags. The DPDK building process is related to host kernel version and ovs need to invoke DPDK functions, that means it's hard to provide an ovs-dpdk docker image for all circumstances. We need to provide the basic install steps docs to manually build and install ovs-dpdk in each circumstance.
2. DPDK need to config lots kernel parameters and even boot flags to enable high performance. That's also strongly related to each circumstance and hard to be containerized.  We also need to provide a basic install and config docs. But I highly recommend user to ask advice from some guy who are professional in DPDK to use it.
3. For hardware requirements, as DPDK will deprive nic control from the system, we need a dedicated nic for ovs-dpdk to transport container network and another nic for normal host network. That's a change to today's kube-ovn network requirement, kube-ovn now doesn't need a dedicated nic.
4. For the code part, it's much easier than the above installation and config steps. We need to change the ovs integration bridge type to `netdev` ,port type to `dpdkvhostuser` and set other ovs-dpdk parameters.

To sum up, we will first try to add the code part to support DPDK type ports and the related installation and config docs to basically support DPDK. However as the installation and config part is complex and varies a lot in different circumstance, users who wants this function may still need lots manual work and validation to really get benefit from it.

**Comment by garyloug** [comment]
Hi all. I’ve been playing around with this over the past couple of weeks and I’ve come to mostly the same conclusions.

I’ve modified the ovs-ovn container to run OVS-DPDK. It works fine but I had the same concerns over portability across different host kernels. If, as you suggest, it’s an option to have the user build their own images locally then at least we have a solution. For me the build takes quite a long time to build DPDK and then build OVS against it. It’s not ideal but maybe it can be sped up in places. The OVS-DPDK install was pretty standard. The pod spec needs hugepages as a resource.

Some things that I don’t think have been mentioned so far:
- Currently kube-ovn puts a veth pair between pod and ovs port. One end of the veth is moved to container network namespace. We can’t do this with OVS-DPDK. When we request a new DPDK port we will end up with something like a vhost-user socket file in a directory such as /var/run/openvswitch/ This can’t be moved between namespaces, it will have to be mounted into the pod just like a regular file. This mounting functionality will be new in kube-ovn.

- The other thing is the K8s default network. I’m almost certain we can’t use OVS-DPDK port for this, K8s api can’t talk to the pod across such an interface and the pod will die. So, we need to continue to provide kernel-OVS as the eth0 default network of the pod, while also providing the OVS-DPDK port - i.e. we need to provide multiple interfaces to the pod. Maybe this should eventually be built into kube-ovn, but for the moment I’ve been testing with Multus to provide the multiple interfaces. Kube-ovn seems to work fine with Multus. 

My next step is kind of a proof of concept. I’m going to test with kube-ovn providing the default network and see if I can get https://github.com/intel/userspace-cni-network-plugin working with my containerized OVS-DPDK to provide me with the DPDK dataplane. This is just a POC, Userspace CNI has the functionality to mount the socket file, as mentioned above.

I’ll also consult with our OVS-DPDK team to see is there anything that can be done with the image portability.

**Comment by oilbeater** [comment]
@garyloug Thanks! I am just trying ovs-dpdk and not go that far. It really helps me a lot.

I'm also investigating Multus and userspace-cni-network-plugin to improve kube-ovn.

I have some questions about the connectivity between the default network and OVS-DPDK network. kube-ovn now use a join subnet to connect container network with host network through ovn0. If port in this subnet still use OVS kernel datapath, can host and ovs-dpdk network be connected.

If possible, can you make a pr to share the modification to kube-ovn and steps to use it, so we can learn from your design and help test if dpdk will influence other existing functions.

**Comment by garyloug** [comment]
Hi @oilbeater , so if I understand your question correctly, you are asking if OVS-DPDK can be used to provide the standard/kernel OVS ports that we need for the K8s default network. Yes it can, OVS-DPDK can still provide kernel networking. You just don't set datapath_type=netdev for a given bridge and it'll use standard kernel interfaces. I think it’s even possible to put the standard kernel interfaces on the netdev dpdk bridge but performance will be bad due to the copy from user to kernel space.

I hope I’ve understood your question correctly.

Happy to share my changes, should be able to create a PR in coming days.

**Comment by oilbeater** [comment]
@garyloug any update on it? We are looking forward to it.

**Comment by garyloug** [comment]
Hi @oilbeater I rebooted my system early last week and my ovs-dpdk stopped working after. I wanted to fix this before sharing anything. It looks like the steps mentioned here have fixed most of my issues: https://mail.openvswitch.org/pipermail/ovs-discuss/2016-September/042605.html

I still haven't managed to get userspace-cni working with the containerized ovs-dpdk, but I'll tidy up my files today and share what I have for now.

**Comment by garyloug** [comment]
I've closed the PR previously mentioned. This was just a proof of concept I shared.
New PR is here: https://github.com/alauda/kube-ovn/pull/352
